### PR TITLE
[homeconnect] 9 channels added for washer appliance

### LINKED ISSUE #10704: [homeconnect] Support of additional washer options not fully covered by the API
My washing machine provides a i-DOS feature which is apparently different to what the binding implemented.
On the machine, for each washing program, I can enable or disable i-DOS1 / i-DOS2.

Strangely, there is no event published when you change it locally :-(

But when you send a GET request /programs/selected, you get two options "LaundryCare.Washer.Option.IDos1.Active" and "LaundryCare.Washer.Option.IDos2.Active" with a boolean value..

So we could add two new channels, at least to display the current state. I don't know yet if the API allows updating these options remotely. Without SSE event to inform of local change of these controls, it will not be obvious to update  the channels. But at least they could be set when a new program is selected.

I will work on this enhancement.

### PR DESCRIPTION
**Pull request by lolodomo**
Corresponding to 9 program options not fully supported by the API.

Fix #10704

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
@bruestel : I chose to reuse the 2 i-Dos channels and to rename the old ones, because it made more sense in my opinion. Old channel "laundry_care_washer_idos1" is now renamed into "laundry_care_washer_idos1_level" and the new one is named "laundry_care_washer_idos1". Of course, it will lead to problems for users having these 2 channels already linked when moving to the new version including this enhancement. If you think it is more important to not annoy users than having consistent names, I could keep "laundry_care_washer_idos1" unchanged and name my new channel "laundry_care_washer_idos1_active" for example.

The added channels are for the washer appliance. I am sure some of them could be added to the washerdryer appliance too. But difficult to be sure without such appliance to verify.

**Comment by lolodomo** [comment]
Just to explain, if some of these new readonly channels (defined as advanced channels) are linked, they will be updated when a new program is selected (locally or remotely) and few seconds after the last local change (in fact when remote control is restored). This is will require a call to an API request to retrieve the different value.
If these channels are not linked, there is no change compared to before, that means no additional API request.
Only the washer appliance is currently concerned.

**Comment by bruestel** [comment]
> I chose to reuse the 2 i-Dos channels and to rename the old ones, because it made more sense in my opinion. Old channel "laundry_care_washer_idos1" is now renamed into "laundry_care_washer_idos1_level" and the new one is named "laundry_care_washer_idos1".

That's fine with me. Makes more sense, as you said. 

> The added channels are for the washer appliance. I am sure some of them could be added to the washerdryer appliance too. But difficult to be sure without such appliance to verify.

I saw your post in the community thread. Hopefully someone will answer. But in general a `washerdryer` usually behaves the same . I think we could change it as well.

File: bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java [hunk 1]
```diff
@@ -96,13 +114,26 @@ protected void configureEventHandlers(Map<String, EventHandler> handlers) {
...
+                event -> getThingChannel(CHANNEL_WASHER_IDOS1_LEVEL).ifPresent(channel -> updateState(channel.getUID(),
                         event.getValue() == null ? UnDefType.UNDEF : new StringType(event.getValue()))));
         handlers.put(EVENT_WASHER_IDOS_2_DOSING_LEVEL,
-                event -> getThingChannel(CHANNEL_WASHER_IDOS2).ifPresent(channel -> updateState(channel.getUID(),
+                event -> getThingChannel(CHANNEL_WASHER_IDOS2_LEVEL).ifPresent(channel -> updateState(channel.getUID(),
                         event.getValue() == null ? UnDefType.UNDEF : new StringType(event.getValue()))));
     }
 
+    @Override
+    protected boolean isLinkedChannelsForUnsupportedFeatures() {
```
Comment (bruestel) [inline bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java:125]: I'm not sure about the method name. Unfortunately, I can't think of a better name.

File: bundles/org.openhab.binding.homeconnect/src/main/resources/OH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -481,6 +490,76 @@
 		</state>
 		<autoUpdatePolicy>veto</autoUpdatePolicy>
 	</channel-type>
+	<channel-type id="laundry_care_washer_idos1" advanced="true">
+		<item-type>Switch</item-type>
+		<label>i-Dos 1</label>
+		<description>This status indicates whether i-Dos 1 is activated for a washer program of the home appliance. (If
+			appliance
+			supports i-Dos)</description>
```
Comment (bruestel) [inline bundles/org.openhab.binding.homeconnect/src/main/resources/OH-INF/thing/thing-types.xml:498]: Formatting looks curious. Can you put in in one line and apply spotless afterwards?

**Review by bruestel — CHANGES_REQUESTED** [review]
Could you please add the new channels to the README as well.

File: bundles/org.openhab.binding.homeconnect/src/main/resources/OH-INF/thing/thing-types.xml (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.homeconnect/src/main/resources/OH-INF/thing/thing-types.xml:498]: Fixed

File: bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java (hunk 1 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java:125]: Maybe `isChannelLinkedToProgramOptionNotFullySupportedByApi` ?

**Comment by lolodomo** [comment]
I will make a last change to be sure the new channels are correctly set when a program becomes active.

File: bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java (hunk 1 above)
Comment (bruestel) [reply bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java:125]: It's long but better. Could you please use it.

Comment (lolodomo) [reply bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/handler/HomeConnectWasherHandler.java:125]: Done

**Comment by bruestel** [comment]
@lolodomo Thanks, your changes are looking good. Can you please also update the README.md file.

**Comment by lolodomo** [comment]
> Thanks, your changes are looking good. Can you please also update the README.md file.

Done.

My change is now finished and you can have a final review.

File: bundles/org.openhab.binding.homeconnect/README.md [hunk 3]
```diff
@@ -56,8 +56,17 @@ After the bridge has been added and authorized, devices are discovered automatic
...
+| laundry_care_washer_idos2_level | String | false | This status defines the i-Dos 2 dosing level of a washer program of the home appliance (if appliance supports i-Dos). | washer | 
+| laundry_care_washer_idos1 | Switch | true | This status indicates whether i-Dos 1 is activated for a washer program of the home appliance. (If appliance supports i-Dos) | washer | 
+| laundry_care_washer_idos2 | Switch | true | This status indicates whether i-Dos 2 is activated for a washer program of the home appliance. (If appliance supports i-Dos) | washer | 
+| laundry_care_washer_vario_perfect | String | true | This status defines the vario perfect mode of a washer program of the home appliance. | washer, washerdryer | 
+| laundry_care_washer_less_ironing | Switch | true | This status indicates whether less ironing is activated for a washer program of the home appliance. | washer, washerdryer | 
+| laundry_care_washer_pre_wash | Switch | true | TThis status indicates whether the pre-wash is activated for a washer program of the home appliance. | washer, washerdryer |
```
Comment (bruestel) [inline bundles/org.openhab.binding.homeconnect/README.md:65]: 'TThis'

Comment (lolodomo) [reply bundles/org.openhab.binding.homeconnect/README.md]: Fixed

**Review by bruestel — APPROVED** [review]

**Review by fwolter — APPROVED** [review]

**Pull request merged**